File: backtracking/prob1863.py
class Solution:
    def subsetXORSum(self, nums: List[int]) -> int:
        
        n = len(nums)

        flags = []
        
        subsets = set()
        
        def walk(flags, i):
            if i == n:
                indices = []
                for i in range(n):
                    if flags[i]:
                        indices.append(i)
                subsets.add(tuple(indices))
                return 
            
            flags.append(False)
            walk(flags, i+1)
            flags.pop()
            flags.append(True)
            walk(flags, i+1)
            flags.pop()
            
        walk(flags, 0)
        
        ans = 0
        
        for t in subsets:
            tmp = 0
            for i in t:
                tmp ^= nums[i]
            ans += tmp
            
        return ans

class Solution:
    def subsetXORSum(self, nums: List[int]) -> int:
        
        n = len(nums)

        flags = []
        subsets = set()
        
        def walk(flags, i):
            if i == n:        
                subsets.add(tuple(flags))
                return 
            
            walk(flags, i+1)
            flags.append(i)
            walk(flags, i+1)
            flags.pop()
            
        walk(flags, 0)
        
        ans = 0
        
        for t in subsets:
            tmp = 0
            for i in t:
                tmp ^= nums[i]
            ans += tmp
            
        return ans

File: math/prob1863.kt
class Solution {

    fun subsetXORSum(nums: IntArray): Int {
        val n = nums.size
        val ans = IntArray(1)
        val l = mutableListOf<Int>()
        walk(l, nums, ans)
        return ans[0]
    }

    fun walk(arr: MutableList<Int>, nums: IntArray, ans: IntArray) {
        val m = arr.size
        val n = nums.size
        if (m == n) {
            ans[0] += helper(arr)
        } else {
            arr.add(0)
            walk(arr, nums, ans)
            // arr.removeLast()
            arr.removeAt(arr.size-1)
            arr.add(nums[m])
            walk(arr, nums, ans)
            arr.removeAt(arr.size-1)
        }
    }

    fun helper(arr: MutableList<Int>): Int {
        val n = arr.size
        var ans = arr.reduce {acc, it -> acc xor it}
        return ans
    }
}
